pkg/client/ssa_client.go

```diff
@@ -0,0 +1,218 @@
+// have its GVK set.
+func SkipIf(test func(client.Object) bool) SsaApplyObjectOption {
+	return func(config *ssaApplyObjectConfiguration) {
+		config.skipIf = test
+	}
+}
+
+// DetermineUpdate instructs the ApplyObject function to truly test whether the object
+// was updated or not during the apply operation. By default, the function always returns
+// true on success which is more efficient and sufficient in most circumstances.
+func DetermineUpdate(value bool) SsaApplyObjectOption {
+	return func(config *ssaApplyObjectConfiguration) {
+		config.determineUpdate = value
+	}
+}
+
+// EnsureLabels makes sure that the provided labels are applied to the object even if
+// the supplied object doesn't have them set.
+func EnsureLabels(labels map[string]string) SsaApplyObjectOption {
+	return func(config *ssaApplyObjectConfiguration) {
+		config.newLabels = labels
+	}
+}
+
+// Configure sets the owner reference and merges the labels. Other options modify the logic
+// of apply function and therefore need to be checked manually.
+func (c *ssaApplyObjectConfiguration) Configure(obj client.Object, s *runtime.Scheme) error {
+	if c.owner != nil {
+		if err := controllerutil.SetControllerReference(c.owner, obj, s); err != nil {
+			return err
+		}
+	}
+	MergeLabels(obj, c.newLabels)
+
+	return nil
+}
+
+// ApplyObject creates the object if is missing and if the owner object is provided, then it's set as a controller reference.
+// If the objects exists then when the spec content has changed (based on the content of the annotation in the original object) then it
+// is automatically updated. If it looks to be same then based on the value of forceUpdate param it updates the object or not.
+// The return boolean says if the object was either created or updated (`true`). If nothing changed (ie, the generation was not
+// incremented by the server), then it returns `false`.
+//
+// NOTE: the return boolean IS ALWAYS TRUE on success by default. This is much more efficient. If you truly need
+// to determine whether the object has been updated or not (which is not the case anywhere in the codebase but the tests),
+// you can use the DetermineUpdate() option.
+func (c *SsaApplyClient) ApplyObject(ctx context.Context, obj client.Object, options ...SsaApplyObjectOption) (bool, error) {
+	config := newSsaApplyObjectConfiguration(options...)
+	if err := config.Configure(obj, c.Client.Scheme()); err != nil {
+		return false, err
+	}
+
+	if err := prepareForSSA(obj, c.Client.Scheme()); err != nil {
+		return false, err
+	}
+
+	if config.skipIf != nil && config.skipIf(obj) {
+		return false, nil
+	}
+
+	updated := true
+
+	// NOTE: once the SSA migration is not needed anymore, read the orig conditionally only when config.determineUpdate is true.
+
+	// this cannot fail - we would have failed previously already because the scheme wouldn't have known the object's kind
+	origRo, _ := c.Client.Scheme().New(obj.GetObjectKind().GroupVersionKind())
```

Comment:
The same thing again - we shouldn't rely on the fact that he object's kind will be always present in the scheme.
I would rather rely on Unstructured type in this case

Reply:
IMHO, the kind needs to be in the scheme even for the unstructured objects. But let me experiment with that.

Reply:
will be fixed